feat(curriculum): working with rdbs lectures (#61450)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/25-front-end-development/lecture-working-with-relational-databases/687ea6c021dd5913cf8fd8dd.md

---
id: 687ea6c021dd5913cf8fd8dd
title: What Are Relational Databases, and How Do They Differ from Non-Relational Databases?
challengeType: 19
dashedName: what-are-relational-databases-and-how-do-they-differ-from-non-relational-databases
---

# --description--

A relational database is a collection of data organized into tables. These tables are made of rows and columns.

Rows represent individual records. For example, if you're storing user data, each row could represent a specific user.

> Columns represent the attributes or fields that describe each record. The table might include columns for the most important user information, including name and national ID number.

Reviewer: ```suggestion
Columns represent the attributes or fields that describe each record. The table might include columns for common user attributes, such as name and national ID number.
```

One of the key characteristics of relational databases is that they allow us to connect different pieces of information by linking tables through common attributes, making it easier to see how they relate to each other.

> Relational databases require a schema. The schema defines the structure and organization of the database, including its tables, columns, data types, relationships, and constraints.

Reviewer: ```suggestion
Relational databases require a schema. The schema defines the overall structure of the database, including its tables, columns, data types, relationships, and constraints.
```

> All tables in a relational database have a primary key, which is a unique identifier for each row. Primary keys are essential for creating relationships between tables through foreign keys. You'll learn more about primary keys, foreign keys, and how they work in the coming lectures. For now, you should know that they are essential for linking records in different tables.

Reviewer: ```suggestion
All tables in a relational database have a primary key, which is a unique identifier for each row. Primary keys are used to uniquely identify each row in a table. They enable relationships between tables via foreign keys. You’ll learn more about how these work in upcoming lectures—for now, just know they are fundamental to linking data across tables.
```

For example, let's say that you're creating a database for an organization that runs a wildlife conservation center for critically endangered animals.

One of the tables could store general information about the animals, you might call it `animals`. Each row could represent an animal, while each column could represent an attribute of that animal, like species, size, weight, age, and other unique characteristics.

> You should assign a unique ID to each animal to identify it in the database. To do this, you would usually add an `id` column. This would be the primary key of the table because it's unique for each animal. The ID of each animal could be used to link its record in this table to related records in other tables.

Reviewer: ```suggestion
To identify each animal uniquely, you would typically add an `id` column to the table. This would be the primary key of the table because it's unique for each animal. The ID of each animal could be used to link its record in this table to related records in other tables.
```

> Another potential table could be a `species` table, where you would store information about all the species in the center, such as their habitat and conservation status.

Reviewer: This section could be expanded slightly to show relationship
```suggestion
You could also have a `species` table to store information about each species, such as their habitat and conservation status. Each animal in the `animals` table could reference a species from this table using a species ID.
```

You could also create a table to store the veterinary records of the animals. This way, the center can track their health, treatments, and medications.

> Once you have this general schema, you could link these tables to the main table, where you store the general information about the animals.

Reviewer: ```suggestion
Once you have this general schema, you could link these tables to the `animals` table, which acts as the central point for animal-related data.
```

> This way, you don't have to keep all the information about an animal in the same table. You can store the data in different tables and link them as necessary. By creating these relationships, you can get the specific data you need with a single query.

Reviewer: ```suggestion
This way, you don't have to keep all the information about an animal in the same table. Instead, you can store related data in separate tables and link them through relationships. By creating these relationships, you can get the specific data you need with a single query.
```

A query is a request for specific data from the database. Examples of queries that you could make in the context of our example would be finding all animals that need a veterinary checkup, finding the most critically endangered species in the center, finding the number of animals who were born in captivity, and more. You can customize the queries to fit your needs.

> Our example was related to wildlife conservation but relational databases have a wide range of applications in many different areas, including healthcare, construction, management, business, gaming, education, government, e-commerce, social media, and more.

Reviewer: ```suggestion
Our example was related to wildlife conservation, but relational databases have a wide range of applications across industries such as healthcare, construction, business, gaming, education, government, e-commerce, social media, and more.
```
